[3/N] Add get_option/set_option function in backend interface

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #11865
* __->__ #11391
* #11390
* #11389



Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
@exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

cccclai committed:
[3/N] Add update function in backend interface

Add update function in backend interface class.

Next step, will add update API in the method and then module

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add update function in backend interface"


Add update function in backend interface class.

Next step, will add update API in the method and then module

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add update function in backend interface"


Add update function in backend interface class.

Next step, will add update API in the method and then module

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add update function in backend interface"


Add update function in backend interface class.

Next step, will add update API in the method and then module

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

[ghstack-poisoned]

kimishpatel commented on runtime/backend/interface.h:
```diff
@@ -99,6 +101,20 @@ class BackendInterface {
+   * @retval Error::Ok if successful.
+   */
+   ET_NODISCARD virtual Error update(
+    BackendUpdateContext& context,
+    const executorch::runtime::ArrayRef<BackendOption>& backend_options) const {
+        return Error::Ok;
+      };
```
what is the implementation of this? If it is in the next diff, can you pull it here?

cccclai replied:
It's the base function. If backends didn't implement this function, this function will be called. If we set it to virtual, then we need all backends to implement this fucntion.

kimishpatel commented on runtime/backend/test/backend_interface_update_test.cpp:
```diff
@@ -0,0 +1,285 @@
+  BackendOptions<5> options;  
+};
+
+TEST_F(BackendInterfaceUpdateTest, HandlesInvalidOption) {
```
There are all these tests but I am not sure if they are meanigfully testing anything? do you need all of them?

cccclai replied:
I think it's helpful to make sure we know each option inside the backend is updated successfully, and they are isolated.

cccclai commented on runtime/backend/interface.h:
```diff
@@ -99,6 +101,20 @@ class BackendInterface {
+   * @param[in] args A list of BackendOptions passed in by users.
+   * @retval Error::Ok if successful.
+   */
+   ET_NODISCARD virtual Error update(
```
I kind of like this API name more, but we can discuss about it.

cccclai committed:
Update on "[3/N] Add update function in backend interface"


Add update function in backend interface class.

Next step, will add update API in the method and then module

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add update function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add update function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]

cccclai committed:
Update on "[3/N] Add get_option/set_option function in backend interface"




Add update function in backend interface class. The update function will receive the backend options from dispatched by the ET runtime.

ET runtime's logic: loop over each backend and it's corresponding backend options, dispatch the backend options to the corresponding backend

Next step, will add update API in the method and then module
exported-using-ghexport

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242/)

Differential Revision: [D75919242](https://our.internmc.facebook.com/intern/diff/D75919242)

[ghstack-poisoned]
